lib/libutil/cpuset.3
.\" Copyright (c) 2017 Baptiste Daroussin <bapt@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd September 08, 2024
.Dt CPUSET 3
.Os
.Sh NAME
.Nm cpuset_parselist,
.Nm domainset_parselist
.Nd utility functions for
.Xr cpuset 2
handling
.Sh LIBRARY
.Lb libutil
.Sh SYNOPSIS
.In sys/cpuset.h
.In libutil.h
.Ft int
.Fn cpuset_parselist "const char *cpu-list" "cpuset_t *mask"
.Ft int
.Fn domainset_parselist "const char *domain-policy" "domainset_t *domain_mask" "int *policyp"
.Sh DESCRIPTION
The
.Fn cpuset_parselist
function parses a
.Va cpu-list
filling the
.Va mask .
.Pp
The
.Va cpu-list
may include numbers separated by '-' for ranges and commas separating individual
numbers.
A special list of
.Dq all
may be specified in which case the list includes all CPUs from the root set.
.Pp
The
.Fn domainset_parselist
function parses a
.Xr domainset 9
memory domain allocation policy
specified by
.Va domain-policy
filling the
.Va domain_mask
and the
.Va policyp.
A valid
.Va domain-policy
is formatted as
.Ar policy:domain-list.
See the
.Ar -n
flag in
.Xr cpuset 1
for a list of valid domain policies.
.Sh RETURN VALUES
Return values can be the following
.Bl -tag -width Er
.It Dv CPUSET_PARSE_OK
The parsing was successful
.It Dv CPUSET_PARSE_ERROR
The
.Va cpu-list
or
.Va domain-policy
format is invalid
.It Dv CPUSET_PARSE_GETAFFINITY
The
.Xr cpuset_getaffinity 2
call has failed
.It Dv CPUSET_PARSE_OUT_OF_RANGE
The number of supported CPUs or NUMA domains has been exceeded.
The maximum number being
.Va CPU_SETSIZE and
.Va DOMAINSET_SETSIZE
respectively.
.It Dv CPUSET_PARSE_GETDOMAIN
The
.Xr cpuset_getdomain 2
call has failed
.El
.Sh SEE ALSO
.Xr cpuset 1 ,
.Xr cpuset 2 ,
.Xr cpuset 9 ,
.Xr domainset 9 ,
.Sh AUTHORS
.An Jeffrey Roberson Aq Mt jeff@FreeBSD.org


lib/libutil/cpuset.c
/*
 * Copyright (c) 2007, 2008 	Jeffrey Roberson <jeff@freebsd.org>
 * All rights reserved.
 *
 * Copyright (c) 2008 Nokia Corporation
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#define _WANT_FREEBSD_BITSET

#include <sys/types.h>
#include <sys/cpuset.h>
#include <sys/domainset.h>

#include <stdlib.h>
#include <string.h>
#include <libutil.h>
#include <ctype.h>

struct numa_policy {
	const char 	*name;
	int		policy;
};

static struct numa_policy policies[] = {
	{ "round-robin", DOMAINSET_POLICY_ROUNDROBIN },
	{ "rr", DOMAINSET_POLICY_ROUNDROBIN },
	{ "first-touch", DOMAINSET_POLICY_FIRSTTOUCH },
	{ "ft", DOMAINSET_POLICY_FIRSTTOUCH },
	{ "prefer", DOMAINSET_POLICY_PREFER },
	{ "interleave", DOMAINSET_POLICY_INTERLEAVE},
	{ "il", DOMAINSET_POLICY_INTERLEAVE},
	{ NULL, DOMAINSET_POLICY_INVALID }
};

static int
parselist(const char *list, struct bitset *mask, int size)
{
	enum { NONE, NUM, DASH } state;
	int lastnum;
	int curnum;
	const char *l;

	state = NONE;
	curnum = lastnum = 0;
	for (l = list; *l != '\0';) {
		if (isdigit(*l)) {
			curnum = atoi(l);
			if (curnum >= size)
				return (CPUSET_PARSE_OUT_OF_RANGE);
			while (isdigit(*l))
				l++;
			switch (state) {
			case NONE:
				lastnum = curnum;
				state = NUM;
				break;
			case DASH:
				for (; lastnum <= curnum; lastnum++)
					BIT_SET(size, lastnum, mask);
				state = NONE;
				break;
			case NUM:
			default:
				goto parserr;
			}
			continue;
		}
		switch (*l) {
		case ',':
			switch (state) {
			case NONE:
				break;
			case NUM:
				BIT_SET(size, curnum, mask);
				state = NONE;
				break;
			case DASH:
				goto parserr;
				break;
			}
			break;
		case '-':
			if (state != NUM)
				goto parserr;
			state = DASH;
			break;
		default:
			goto parserr;
		}
		l++;
	}
	switch (state) {
		case NONE:
			break;
		case NUM:
			BIT_SET(size, curnum, mask);
			break;
		case DASH:
			goto parserr;
	}
	return (CPUSET_PARSE_OK);
parserr:
	return (CPUSET_PARSE_ERROR);
}

int
domainset_parselist(const char *list, domainset_t *mask, int *policyp)
{
	domainset_t rootmask;
	struct numa_policy *policy;
	const char *l;
	int p;

	/*
	 * Use the rootset's policy as the default for unspecified policies.
	 */
	if (cpuset_getdomain(CPU_LEVEL_ROOT, CPU_WHICH_PID, -1,
	    sizeof(rootmask), &rootmask, &p) != 0)
		return (CPUSET_PARSE_GETDOMAIN);

	if (list == NULL || strcasecmp(list, "all") == 0 || *list == '\0') {
		*policyp = p;
		DOMAINSET_COPY(&rootmask, mask);
		return (CPUSET_PARSE_OK);
	}


	l = list;
	for (policy = &policies[0]; policy->name != NULL; policy++) {
		if (strncasecmp(l, policy->name, strlen(policy->name)) == 0) {
			p = policy->policy;
			l += strlen(policy->name);
			if (*l != ':' && *l != '\0')
				return (CPUSET_PARSE_ERROR);
			if (*l == ':')
				l++;
			break;
		}
	}
	*policyp = p;

	return (parselist(l, (struct bitset *)mask, DOMAINSET_SETSIZE));
}

int
cpuset_parselist(const char *list, cpuset_t *mask)
{
	if (strcasecmp(list, "all") == 0) {
		if (cpuset_getaffinity(CPU_LEVEL_ROOT, CPU_WHICH_PID, -1,
		    sizeof(*mask), mask) != 0)
			return (CPUSET_PARSE_GETAFFINITY);
		return (CPUSET_PARSE_OK);
	}

	return (parselist(list, (struct bitset *)mask, CPU_SETSIZE));
}
